sphinx/util/inspect.py

```diff
@@ -61,6 +61,8 @@ def unwrap_all(obj: Any, *, stop: Callable = None) -> Any:
         elif ispartial(obj):
             obj = obj.func
         elif inspect.isroutine(obj) and hasattr(obj, '__wrapped__'):
+            if hasattr(obj.__wrapped__, '__call__'):
+                return obj  # Don't unwrap wrapped callable types
```

Comment:
```suggestion
            # Don't unwrap wrapped callable types:
            if hasattr(obj.__wrapped__, '__call__'):
                # From Python 3.10, staticmethod and classmethod have
                # "__wrapped__" and "__wrapped__.__call__" attributes.
                # Check that we are in Python 3.9 or earlier, or that the object
                # isn't a staticmethod or classmethod before returning
                # xref: https://docs.python.org/3.10/whatsnew/3.10.html#other-language-changes
                if sys.version_info[:2] < (3, 10):
                    return obj
                if not (isstaticmethod(obj) or isclassmethod(obj)):
                    return obj
```

A

Reply:
So this seems pretty unpleasant and complicated. I guess at this point I would back up and note that after all, Sphinx must be using its own `isfunction` implementation for a reason (i.e. some difference from stdlib `isfunction` must actually be desired). I'm happy to commit this suggestion if you do think making a change is best, but OTOH if the current behavior is defensible then a custom Documenter subclass to override does work fine.